### ISSUE TITLE
Question: BigQuery example with nested data

### ISSUE THREAD

**Issue by kurtmaile** [GoogleCloudPlatform/google-cloud-go#905]
Hi,

The bigquery samples show a very simple, 1 level example using stackoverflow struct that has 2 top level fields.

I am having some issues with something that is more realistic and nested a few levels (Im not getting an error message back to indicate what it is, but I suspect my valueSaver 'Save()' is not implemented correctly.  My data model is non-trivial with nested structs and arrays of structs, and I think this is where my issue is (can get it to work for a simple top level example) - but its when its more complex I am struggling.  Namely:

1. How do I make a nested BQ record nullable?  There are explicity types for the the primitive types and also an example, but my data has nested, optional struct.  My BQ table gets generated with such a struct as 'required', when needs to be nullable.  Also what about an array of struct?

2. Is the valueSaver interface impelmentation for structs still required?  I believe it is looking at code (but maybe being phased out in next major release?), bit of an overhead to need to do this but I guess its required for now right?

So in that case, what would does the value saver Save() method need to look like with a root struct that has a nested struct and an array of struct?

Basically Im looking for a non-trivial example that isnt 1 top level, has a child struct and array of struct, both need to be nullable.  Sorry to bother on this but I cant find an example anywhere!

Thanks heaps!
Kurt

**Comment by kurtmaile** [comment]
Can anyone help on this please?

**Comment by jadekler** [comment]
Hey @kurtmaile - sorry for the lag getting to your question. We're understaffed this week, but next week should be back up to strength. We'll hopefully have an answer to you soon!

**Comment by kurtmaile** [comment]
Hey no problem at all, thanks for letting me know. Excited to use the library for BQ is all but need some clarity (via example) on a non-trivial nested struct example as described and how to set the value saver up.  e.g below.....I would want both D and E to be nullable as well.  Thanks!!

```
type A struct {
	B string              `bigquery:"b"`
	C bigquery.NullString `bigquery:"c"`
	D *X                  `bigquery:"d"`
	E []*X                `bigquery:"e"`
}

type X struct {
	Y bigquery.NullString `bigquery:"y"`
	Z *Z     `bigquery:"z"`
}

type Z struct {
	W bigquery.NullString `bigquery:"w"`
}
```

**Comment by kurtmaile** [comment]
Hi guys,  Can anyone help with this please?  Much appreciated :)

**Comment by jba** [comment]
You don't need to write a custom ValueSaver for that struct. All you need to do, I think, is change the tag on `D` to read `bigquery:"d,nullable"`. The `E` field is repeated, so it is already nullable.

Here's some code that will display the inferred schema in a nice way, so you can check that it's what you want. Just add your struct definitions to the same file and run.
```
func main() {
    schema, err := bigquery.InferSchema(A{})
    if err != nil {
        log.Fatal(err)
    }
    displaySchema(schema, 0)
}

func displaySchema(s bigquery.Schema, level int) {
    for _, f := range s {
        var m string
        switch {
        case f.Repeated:
            m = "REPEATED"
        case f.Required:
            m = "REQUIRED"
        default:
            m = "OPTIONAL"
        }
        for i := 0; i < level; i++ {
            fmt.Print("    ")
        }
        fmt.Printf("%s\t%s\t%s\n", f.Name, f.Type, m)
        if f.Type == bigquery.RecordFieldType {
            displaySchema(f.Schema, level+1)
        }
    }
}
```
After changing the struct tag on `D` as described above, the output is:
```
b       STRING  REQUIRED
c       STRING  OPTIONAL
d       RECORD  OPTIONAL
    y   STRING  OPTIONAL
    z   RECORD  REQUIRED
        w       STRING  OPTIONAL
e       RECORD  REPEATED
    y   STRING  OPTIONAL
    z   RECORD  REQUIRED
        w       STRING  OPTIONAL
```

**Comment by kurtmaile** [comment]
Hi Jonathan,  Thanks heaps will try this out soon.  So to be clear, am I still required to implement Save() from the ValueSaver interface?  You said Im not required to write a 'custom' ValueSaver method but not sure what 'custom' entails......Say for example attribute 'b' (a key) was what I wanted the BQ deduping field to be on the streaming inserts, this is what was defined in the 'ValueSaver.Save()' func, but I dont want to implement Save at all if I dont have.  Would having this extra deduping requirement mean I need to impelment 'Save()' for my struct?  If so, do you only need to do the top level struct?  Sorry for the questions, just trying to get clarity on the best (and least amount of boiler plate code) for my requirement for a complex data model.  Thanks heaps! :)

**Comment by jba** [comment]
If you want to use your own insert ID, then you can use a `StructSaver`. See the example at https://godoc.org/cloud.google.com/go/bigquery#Uploader.

**Comment by kurtmaile** [comment]
Thanks heaps Jonathan!  This seems like its what I need - table schema works exactly, just trying to insert (put) a struct into the table though and its failing. The returned error message though isnt giving me a clue to the issue:

'1 row insertion failed'

Is there more detail elsewhere in the errors returned?  Ill keep investigating anyway - my BQ table schema is spot on at least.

Really appreciate the time to help, Im close to getting the pattern down using this within AWS lambda to publish data to BQ.  Cant wait for native go support in google cloud functions!! Much appreciated.

Cheers mate
Kurt

**Comment by kurtmaile** [comment]
Ok think I found the issue.  My custom time partitioning column is a Time.time, but it is in the time.RFC3339Nano format.  Think this is causing the issue trying to either convert it to bigquery.NullTimestamp OR in BQ itself on insert.  Need to investigate further

Thanks heaps for all your help on this! Really great and helpful golang community

**Comment by kurtmaile** [comment]
Hi Jonathan,  

Sorry me again.  We use RFC3339Nano as a default time format and is standard with go time library.  

This actually comes into go in a string format (from json payload).  I am trying to use a field that is of this format as the custom time partitioning field (TimePartitoning.Field) on a BQ date partitoned table

However I seem to be having issues with inserting a row (StructSaver) using this format.  

Note: if I use time.Now() as a benchmark test for partitoning field, the data write to BQ fine.....sweet! :)  My complex data makes it into BQ.

But, what I really need to do is to convert a date string in nano format, to a time.Time, and use that.  Now for the life of me I cant get this to work - no matter how I try format the date.  As an example:

Say you have a nano date string to use as a Time.time for custom partitioning

```
dateString := "2018-02-14T15:28:18.165413538Z"
t := time.Parse(time.RFC3339Nano, dateString)
```

and then use ```t``` as the partiton field gives me an error - ```1 row insertion failed```

If I try then 'dropping' the nanos (format to time.RFC3339 then parse again) this also fails - I thought maybe try dropping the nanos precision shown in a funciton below and use the return time.Time as the partition value....but still the same error!  Argg......

```
func ParseDateTimeDropNanos(nanoDate string) time.Time {
	dt := time.Parse(time.RFC3339Nano, dateString)
	nonNanos := dt.Format(time.RFC3339)
	return time.Parse(nonNanos, time.RFC3339)
}
```

As mentioned, just using time.Now() shows this works and data is written to BQ, but something about the nano format, and any conversions, that isnt being liked somewhere (maybe BQ?).

Are you able to verify this scenario doesnt work and possible suggest a way to make it work?  Any pro advice most welcome!

Thanks!
Kurt

**Comment by jba** [comment]
I'm confused. I probably don't understand BQ time partitioning as well as I should. You said

> This actually comes into go in a string format (from json payload). I am trying to use a field that is of this format as the custom time partitioning field (TimePartitoning.Field) on a BQ date partitoned table.

Now `TimePartioning.Field` is the name of a BQ column, which must be either a TIMESTAMP or a DATE (BQ types, not Go types). What is the type of your column?

I take it you have a Go struct with a field that corresponds to the BQ partitioning field. Can you show me the definition of that Go field?

**Comment by kurtmaile** [comment]
Hi Jonathan,

Thansk for replying.

My struct yes has a top level field of type 'Pt' of time time.Time.  I also tried it as bigquery.NullTimestamp while testing and still an error.  Below is an example of a struct and creating a mock event where 'Pt' is the top level partitioning field.  The mock uses a nano precision timestamp string and the conversion function above.  Essentially Im trying to create a time.Time.

As mentioned, just using time.Now() as the mock input to 'Pt' works.  Data is written to BQ.  But when my source is a nano string, , and I try converting, it doesnt - even though I produce a valid time.Time instance. 

For clarity on the last post, it is ingested into go program first as a string from json payload, then obviously converted to a time.Time to populate pt in the  BQ struct world for writting to BQ as a proper type.  Below is the canonical type for BQ, it can go to microseconds, not nano seconds.  This makes me think there is  an issue in conversion (trying to reduce the precision) as I do create a valid time.Time.

Thoughts?  Thanks heaps!

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp-type

  

```
type BQOrderBusinessEvent struct {
	Id         string                 `bigquery:"id"`
	Event      *BQEvent               `bigquery:"event,nullable"`
	Pt         time.Time `bigquery:"pt"`
}

func GetMockBQOrderBusinessEvent_Minimum() *BQOrderBusinessEvent {
	
	pt := util.ParseDateTimeDropNanos("2018-02-14T15:28:18.165413538Z")
	fmt.Printf("%v", pt)
	ev := &BQOrderBusinessEvent{
		Id: crypto.NewUUID().String(),
		Event: &BQEvent{
			EventName:          "OrderAccepted",
			EventCategory:      "Order",
			EventOccurredTime:  "2018-02-14T15:28:18.165413538Z",
			EventReceivedTime:  "2018-02-14T15:28:18.165413538Z",
			EventProcessedTime: "2018-02-14T15:28:18.165413538Z",
		},
		Pt: pt,
	}
	return ev
}
```

**Comment by jba** [comment]
Let me try to duplicate what you've done and get back to you.

**Comment by kurtmaile** [comment]
Awesome thanks Jonathan!

**Comment by jba** [comment]
Reproduced the problem. There is more information available in the error returned from `Uploader.Put`. Here's how to get it:
```
err := u.Put(ctx, ev)
if pme, ok := err.(bigquery.PutMultiError); ok {
    for _, e := range pme {
        fmt.Println(e)
   }
}
```
When you do that, you'll see the message "Value 1518622098165413 for field pt of the destination table XYZ is outside the allowed bounds. You can only stream to date range within 7 days in the past and 3 days in the future relative to the current date."

That's actually documented [here](https://cloud.google.com/bigquery/docs/partitioned-tables#comparing_partitioning_options), in the table row beginning "Streaming inserts".

**Comment by kurtmaile** [comment]
Hi Jonathan,

This is super helpful thanks. Really really appreciate all your help, looking forward to getting on building with it :)

All the best and thanks once again
Kurt
